edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dtmanager
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/constants"
+	pb "github.com/kubeedge/kubeedge/edge/pkg/apis/dmi/v1"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dmiclient"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dmiserver"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
+	"github.com/kubeedge/kubeedge/edge/pkg/metamanager/dao"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+)
+
+//TwinWorker deal twin event
+type DMIWorker struct {
+	Worker
+	Group string
+}
+
+var (
+	//dmiActionCallBack map for action to callback
+	dmiActionCallBack map[string]CallBack
+
+	mapperList      map[string]*pb.MapperInfo
+	deviceModelList map[string]*v1alpha2.DeviceModel
+	deviceList      map[string]*v1alpha2.Device
+)
+
+//Start worker
+func (dw DMIWorker) Start() {
+	klog.Infoln("dmi worker start")
+	initDMIActionCallBack()
+	initDeviceModelInfoFromDB()
+	initDeviceInfoFromDB()
+	initDeviceMapperInfoFromDB()
+
+	go dmiserver.StartDMIServer(mapperList, deviceList, deviceModelList)
+
+	for {
+		select {
+		case msg, ok := <-dw.ReceiverChan:
+			if !ok {
+				return
+			}
+
+			if dtMsg, isDTMessage := msg.(*dttype.DTMessage); isDTMessage {
+				if fn, exist := dmiActionCallBack[dtMsg.Action]; exist {
+					err := fn(dw.DTContexts, dtMsg.Identity, dtMsg.Msg)
+					if err != nil {
+						klog.Errorf("DMIModule deal %s event failed: %v", dtMsg.Action, err)
+					}
+				} else {
+					klog.Errorf("DMIModule deal %s event failed, not found callback", dtMsg.Action)
+				}
+			}
+
+		case v, ok := <-dw.HeartBeatChan:
+			if !ok {
+				return
+			}
+			if err := dw.DTContexts.HeartBeat(dw.Group, v); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func initDMIActionCallBack() {
+	dmiActionCallBack = make(map[string]CallBack)
+	dmiActionCallBack[dtcommon.MetaDeviceOperation] = dealMetaDeviceOperation
+}
+
+func dealMetaDeviceOperation(context *dtcontext.DTContext, resource string, msg interface{}) error {
+	message, ok := msg.(*model.Message)
+	if !ok {
+		return errors.New("msg not Message type")
+	}
+	resources := strings.Split(message.Router.Resource, "/")
+	if len(resources) != 3 {
+		return fmt.Errorf("wrong resources %s", message.Router.Resource)
+	}
+	var device v1alpha2.Device
+	var dm v1alpha2.DeviceModel
+	switch resources[1] {
+	case constants.ResourceTypeDevice:
+		switch message.GetOperation() {
+		case model.InsertOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcCreateDevice(&device)
```

Comment:
I think we can remove grpc prefix from the func name

Reply:
done